core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ZooCache.java

```diff
@@ -316,43 +277,46 @@ public List<String> getChildren(final String zPath) {
       public List<String> run() throws KeeperException, InterruptedException {
 
         var zcNode = nodeCache.get(zPath);
-        if (zcNode != null && zcNode.childrenSet) {
+        if (zcNode != null && zcNode.cachedChildren()) {
           return zcNode.getChildren();
         }
 
         log.trace("{} {} was not in children cache, looking up in zookeeper", cacheId, zPath);
 
-        try {
-          zcNode = nodeCache.compute(zPath, (zp, zcn) -> {
-            // recheck the children now that lock is held on key
-            if (zcn != null && zcn.childrenSet) {
-              return zcn;
-            }
+        zcNode = nodeCache.compute(zPath, (zp, zcn) -> {
+          // recheck the children now that lock is held on key
+          if (zcn != null && zcn.cachedChildren()) {
+            return zcn;
+          }
 
-            try {
-              final ZooKeeper zooKeeper = getZooKeeper();
-              List<String> children;
-              children = zooKeeper.getChildren(zPath, watcher);
-              if (children != null) {
-                children = List.copyOf(children);
-              }
-              return new ZcNode(zcn, children);
-            } catch (KeeperException e) {
-              throw new ZcException(e);
-            } catch (InterruptedException e) {
-              throw new ZcInterruptedException(e);
+          try {
+            final ZooKeeper zooKeeper = getZooKeeper();
+            // Register a watcher on the node to monitor creation/deletion events for the node. It
+            // is possible that an event from this watch could trigger prior to calling getChildren.
+            // That is ok because the compute() call on the map has a lock and processing the event
+            // will block until compute() returns. After compute() returns the event processing
+            // would clear the map entry.
+            Stat stat = zooKeeper.exists(zPath, watcher);
```

Comment:
I'm wondering if this is going to cause an increase in the number of watches that we cannot clean up (or have a hard time doing so). If the node doesn't exist, and never comes into existence, the watch will be registered forever. This was one of the problems that motivated moving the system/table/namespace configs to a single node, because we were watching for exists on property keys for properties that were never stored in ZK, or for tables that had already been deleted, and so the properties would never come into being.

So, caching non-existence seems risky... prone to the same sort of memory leak, where the ZK client memory keeps growing and growing due to an unbounded number of watches on nodes that don't exist and may never exist.

Reply:
That is a good point.  Makes me wonder about the existing functionality of  caching non-existence when get node data.

The use case I was trying to solve was looking up tablet server locks.  For this use case if we have AC="num accumuo clients reading from tservers", TRG="number of tablet severs resource groups", ATPC="average number of tservers accessed per accumulo client" then the number of watches for non-existence created would be AC*ATPC*(TRG-1).  If any two of those three numbers are large then it could result in a lot of watches.

I made #5134 a blocker for 4.0.0.  Thinking prior to releasing 4.0 it would be good to get a handle on what the watches look like.

Looking into what zookeper might offer I discovered [WatchManager](https://zookeeper.apache.org/doc/current/apidocs/zookeeper-server/org/apache/zookeeper/server/watch/WatchManager.html) which is new to me.  Going to look into this a bit and see if there anything that can be done with it in zoocache to avoid excessive watches.

Reply:
In ee80c70 updated to remove entries in zoocache that have not been accessed in 3 minutes.  This also removes their watches. I think this covers a lot of the issues of creating watches that hang around forever.  Looking into this I think the following follow on work could be done to improve the watch situation.

 * When something is removed from zoocache also remove the watch.  Currently when something is removed from zoocache explicitly its watch is just left there. 
 * Only set a watch on things that are accesses more than once. I suspect there are things in zoocache that are only accessed once and a watch is set.   These expire changes will help with this.  We could reduce watches further in a follow on by only setting them on the 2nd access.

Reply:
Also found WatchManager is not public API for ZK.  Fortunately a subset of its functionality is available in the public API.

Reply:
I'm not sure when the removeWatches functionality was added, but I believe it was relatively recent. It might be newer than the oldest ZK we support, and we might need to document that we require a newer version... I'm not sure.

Reply:
Zookeeper.removeWatches has a since 3.5.0 tag in its javadoc.

Reply:
I was able to build this branch using zookeeper 3.7.1 and `Zookeeper.removeWatches` was present. 

The [release notes for 2.1.0](https://accumulo.apache.org/release/accumulo-2.1.0/)  mention 3.8.0 so this change should be fine. 
```
During much of this release’s development, ZooKeeper 3.5 was used as a minimum. However, that version reach its end-of-life during development, and we do not recommend using end-of-life versions of ZooKeeper. The latest bugfix version of 3.6, 3.7, or 3.8 should also work fine. By default, our POM depends on 3.8.0.
```